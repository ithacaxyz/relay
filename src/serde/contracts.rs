//! Helpers for serialization and deserialization of entrypoint and delegation address pairs.

use crate::config::EntryWithDelegation;
use alloy::primitives::Address;
use eyre::WrapErr;
use std::str::FromStr;

/// Parse a string in the format "0xENTRYPOINT,0xDELEGATION" into an [`EntryWithDelegation`].
pub fn parse_entrypoint_with_delegation(s: &str) -> eyre::Result<EntryWithDelegation> {
    let parts: Vec<&str> = s.split(',').collect();
    if parts.len() != 2 {
        return Err(eyre::eyre!("Expected format ENTRYPOINT,DELEGATION"));
    }
    let entrypoint = Address::from_str(parts[0]).wrap_err("Entrypoint address parse failed.")?;
    let delegation = Address::from_str(parts[1]).wrap_err("Delegation address parse failed.")?;

    Ok(EntryWithDelegation { entrypoint, delegation })
}
