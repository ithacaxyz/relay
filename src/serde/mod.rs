//! Serde helpers.

pub mod duration;
pub mod fn_selector;
pub mod hash_map;
pub mod key_role;
pub mod timestamp;
pub mod trace_context;
