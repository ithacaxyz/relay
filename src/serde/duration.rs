//! Helpers for serializing and deserializing [`Duration`].

use serde::{self, Deserialize, Deserializer, Serializer};
use std::time::Duration;

/// Serializes [`Duration`] as seconds.
pub fn serialize<S>(duration: &Duration, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    let duration = duration.as_secs();
    serializer.serialize_u64(duration)
}

/// Deserializes seconds into a [`Duration`].
pub fn deserialize<'de, D>(deserializer: D) -> Result<Duration, D::Error>
where
    D: Deserializer<'de>,
{
    let duration = u64::deserialize(deserializer)?;
    Ok(Duration::from_secs(duration))
}
