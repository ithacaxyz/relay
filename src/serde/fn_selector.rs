//! Helpers for deserializing function selectors.

use std::str::FromStr;

use alloy::{json_abi::Function, primitives::FixedBytes};
use serde::{Deserialize, Deserializer};

/// Deserialize a function selector from either a string or a valid 4-byte array.
///
/// See [`Function::parse`].
pub fn deserialize<'de, D>(deserializer: D) -> Result<FixedBytes<4>, D::Error>
where
    D: Deserializer<'de>,
{
    let s: &str = Deserialize::deserialize(deserializer)?;

    FixedBytes::<4>::from_str(s)
        .or_else(|_| Function::parse(s).map_err(serde::de::Error::custom).map(|f| f.selector()))
}
