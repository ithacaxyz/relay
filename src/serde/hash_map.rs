//! [`toml`] does not allow non-[`String`] keys for [`HashMap`], so we're working around this with a
//! helepr that (de)serializes keys as strings via [`ToString`] and [`FromStr`] implementations.

use alloy::primitives::map::HashMap;
use serde::{self, Deserialize, Deserializer, Serialize, Serializer, de::Error};
use std::{fmt::Display, hash::Hash, str::FromStr};

/// Serializes [`Duration`] as seconds.
pub fn serialize<S, K, V>(map: &HashMap<K, V>, serializer: S) -> Result<S::Ok, S::Error>
where
    K: ToString + Hash + Eq,
    V: Serialize,
    S: Serializer,
{
    map.iter().map(|(k, v)| (k.to_string(), v)).collect::<HashMap<_, _>>().serialize(serializer)
}

/// Deserializes seconds into a [`Duration`].
pub fn deserialize<'de, K, V, D>(deserializer: D) -> Result<HashMap<K, V>, D::Error>
where
    K: FromStr<Err: Display> + Hash + Eq,
    V: Deserialize<'de>,
    D: Deserializer<'de>,
{
    HashMap::<String, V>::deserialize(deserializer).and_then(|map| {
        map.into_iter()
            .map(|(k, v)| K::from_str(&k).map(|k| (k, v)).map_err(D::Error::custom))
            .collect()
    })
}
