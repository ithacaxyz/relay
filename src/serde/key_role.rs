//! Helpers for serialization and deserialization of key roles.

use serde::{Deserialize, Deserializer, Serializer};

/// Deserialize a string into a boolean value.
///
/// - `"admin"` -> `true`
/// - `"normal"` -> `false`
/// - `"session"` -> `false`
pub fn deserialize<'de, D>(deserializer: D) -> Result<bool, D::Error>
where
    D: Deserializer<'de>,
{
    let s: &str = Deserialize::deserialize(deserializer)?;
    match s {
        "admin" => Ok(true),
        "normal" | "session" => Ok(false),
        _ => Err(serde::de::Error::unknown_variant(s, &["admin", "normal", "session"])),
    }
}

/// Serialize a boolean value as `"admin"` (`true`) or `"normal"` (`false`).
pub fn serialize<S>(value: &bool, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    match value {
        true => serializer.serialize_str("admin"),
        false => serializer.serialize_str("normal"),
    }
}
