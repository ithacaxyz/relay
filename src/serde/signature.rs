//! Helpers for serializing and deserializing signatures.

use alloy::{hex, primitives::PrimitiveSignature};
use serde::{Deserialize, Serializer};
use std::str::FromStr;

/// Serializes [`PrimitiveSignature`] as string.
pub fn serialize<S>(signature: &PrimitiveSignature, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    serializer.serialize_str(&hex::encode(signature.as_bytes()))
}

/// Deserialize a [`PrimitiveSignature`] from a String.
pub fn deserialize<'de, D>(deserializer: D) -> Result<PrimitiveSignature, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let s = String::deserialize(deserializer)?;
    PrimitiveSignature::from_str(&s).map_err(serde::de::Error::custom)
}
