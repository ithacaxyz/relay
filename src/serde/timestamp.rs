//! Helpers for serializing and deserializing [`SystemTime`].

use serde::{self, Deserialize, Deserializer, Serializer};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

/// Serializes [`SystemTime`] as the number of seconds since [`UNIX_EPOCH`].
pub fn serialize<S>(time: &SystemTime, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    let timestamp = time.duration_since(UNIX_EPOCH).unwrap_or(Duration::from_secs(0)).as_secs();
    serializer.serialize_u64(timestamp)
}

/// Deserializes a number of seconds since [`UNIX_EPOCH`] into a [`SystemTime`].
pub fn deserialize<'de, D>(deserializer: D) -> Result<SystemTime, D::Error>
where
    D: Deserializer<'de>,
{
    let timestamp = u64::deserialize(deserializer)?;
    Ok(UNIX_EPOCH + Duration::from_secs(timestamp))
}
