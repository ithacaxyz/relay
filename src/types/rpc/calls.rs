//! RPC calls-related request and response types.

use super::{AuthorizeKey, AuthorizeKeyResponse, Meta, RevokeKey};
use crate::{
    error::{IntentError, RelayError},
    types::{
        Account, Asset, AssetDiffs, Call, CreatableAccount, DEFAULT_SEQUENCE_KEY, IERC20, Key,
        KeyType, MULTICHAIN_NONCE_PREFIX_U192, SignedCall, SignedCalls, SignedQuotes,
    },
};
use alloy::{
    consensus::Eip658Value,
    dyn_abi::TypedData,
    primitives::{
        Address, B256, BlockHash, BlockNumber, Bytes, ChainId, TxHash, U256,
        aliases::{B192, U192},
        wrap_fixed_bytes,
    },
    providers::DynProvider,
    rpc::types::Log,
    sol_types::{SolCall, SolEvent},
    uint,
};
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

wrap_fixed_bytes! {
    /// An identifier for a call bundle.
    ///
    /// This is a unique identifier for a call bundle, which is used to track the status of the bundle.
    ///
    /// Clients should treat this as an opaque value and not attempt to parse it.
    pub struct BundleId<32>;
}

/// Key that will be used to sign the call bundle.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallKey {
    /// Type of key.
    #[serde(rename = "type")]
    pub key_type: KeyType,
    /// Public key in encoded form.
    pub public_key: Bytes,
    /// Whether the digest will be prehashed by the key.
    pub prehash: bool,
}

impl CallKey {
    /// The key hash.
    ///
    /// The hash is computed as `keccak256(abi.encode(key.keyType, keccak256(key.publicKey)))`.
    pub fn key_hash(&self) -> B256 {
        Key::hash(self.key_type, &self.public_key)
    }
}

/// Request parameters for `wallet_prepareCalls`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareCallsParameters {
    /// Call bundle to prepare.
    pub calls: Vec<Call>,
    /// Target chain ID.
    #[serde(with = "alloy::serde::quantity")]
    pub chain_id: ChainId,
    /// Address of the account to prepare the call bundle for. It can only be None, if we are
    /// handling a precall
    pub from: Option<Address>,
    /// Request capabilities.
    pub capabilities: PrepareCallsCapabilities,
    /// Key that will be used to sign the call bundle. It can only be None, if we are handling a
    /// precall.
    #[serde(default)]
    pub key: Option<CallKey>,
    /// Required funds on the target chain.
    pub required_funds: Vec<(Address, U256)>,
}

impl PrepareCallsParameters {
    /// Ensures there are only whitelisted calls in precalls and that any upgrade delegation request
    /// contains the latest delegation address.
    pub fn check_calls(&self, latest_delegation: Address) -> Result<(), RelayError> {
        let has_unallowed = |calls: &[Call]| -> Result<bool, RelayError> {
            for call in calls {
                if !call.is_whitelisted_precall(self.from.unwrap_or_default(), latest_delegation)? {
                    return Ok(true);
                }
            }
            Ok(false)
        };

        if self.capabilities.pre_call {
            // Ensure this precall request only has valid calls
            if has_unallowed(&self.calls)? {
                return Err(IntentError::UnallowedPreCall.into());
            }
        } else {
            // Ensure that if the intent is upgrading its delegation, it's to the latest one.
            for call in &self.calls {
                call.ensure_valid_upgrade(latest_delegation)?;
            }
        }

        // Ensure precalls only have valid calls
        for precall in &self.capabilities.pre_calls {
            if has_unallowed(&precall.calls().map_err(RelayError::from)?)? {
                return Err(IntentError::UnallowedPreCall.into());
            }
        }

        Ok(())
    }

    /// Retrieves the appropriate nonce for the request, following this order:
    ///
    /// 1. If `capabilities.meta.nonce` is set, return it directly.
    /// 2. If this is a precall, generate a random sequence key without the multichain prefix and
    ///    return its 0th nonce.
    /// 3. If this is a intent and there are any previous precall entries with the
    ///    `DEFAULT_SEQUENCE_KEY`, take the highest nonce and increment it by 1.
    /// 4. If this is the intent of a non delegated account (`maybe_stored`), return random.
    /// 5. If none of the above match, query for the next account nonce onchain (for
    ///    `DEFAULT_SEQUENCE_KEY`).
    pub async fn get_nonce(
        &self,
        maybe_stored: Option<&CreatableAccount>,
        provider: &DynProvider,
    ) -> Result<U256, RelayError> {
        // Create a random sequence key.
        let random_nonce = loop {
            let sequence_key = U192::from_be_bytes(B192::random().into());
            if sequence_key >> 176 != MULTICHAIN_NONCE_PREFIX_U192 {
                break U256::from(sequence_key) << 64;
            }
        };

        if let Some(nonce) = self.capabilities.meta.nonce {
            Ok(nonce)
        } else if self.capabilities.pre_call {
            Ok(random_nonce)
        } else if let Some(precall) = self
            .capabilities
            .pre_calls
            .iter()
            .filter(|precall| (precall.nonce >> 64) == U256::from(DEFAULT_SEQUENCE_KEY))
            .max_by_key(|precall| precall.nonce)
        {
            Ok(precall.nonce + uint!(1_U256))
        } else if maybe_stored.is_some() {
            Ok(random_nonce)
        } else {
            let eoa = self.from.ok_or(IntentError::MissingSender)?;
            Account::new(eoa, &provider).get_nonce().await.map_err(RelayError::from)
        }
    }

    /// Return a self that will be used as a funding intent
    pub fn build_funding_intent(
        eoa: Address,
        chain_id: ChainId,
        funding_asset: Asset,
        amount: U256,
        fee_token: Address,
        request_key: CallKey,
    ) -> Self {
        // todo: escrow
        let escrow = Address::ZERO;
        let escrow_call = if funding_asset.is_native() {
            Call { to: escrow, value: amount, data: Default::default() }
        } else {
            Call {
                to: funding_asset.address(),
                value: U256::ZERO,
                data: IERC20::transferCall { to: escrow, amount }.abi_encode().into(),
            }
        };

        PrepareCallsParameters {
            calls: vec![escrow_call],
            chain_id,
            from: Some(eoa),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: vec![],
                meta: Meta { fee_payer: None, fee_token, nonce: None },
                revoke_keys: vec![],
                pre_calls: vec![],
                pre_call: false,
            },
            key: Some(request_key),
            required_funds: vec![],
        }
    }
}

/// Capabilities for `wallet_prepareCalls` request.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareCallsCapabilities {
    /// Keys to authorize on the account.
    #[serde(default)]
    pub authorize_keys: Vec<AuthorizeKey>,
    /// Extra request values.
    pub meta: Meta,
    /// Keys to revoke from the account.
    #[serde(default)]
    pub revoke_keys: Vec<RevokeKey>,
    /// Optional preCalls to execute before signature verification.
    ///
    /// See [`Intent::encodedPreCalls`].
    #[serde(default)]
    pub pre_calls: Vec<SignedCall>,
    /// Whether the call bundle is to be considered a precall.
    #[serde(default)]
    pub pre_call: bool,
}

/// Capabilities for `wallet_prepareCalls` response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareCallsResponseCapabilities {
    /// Keys that were authorized on the account.
    #[serde(default)]
    pub authorize_keys: Vec<AuthorizeKeyResponse>,
    /// Keys that were revoked from the account.
    #[serde(default)]
    pub revoke_keys: Vec<RevokeKey>,
    /// The [`AssetDiff`] of the prepared call bundle.
    pub asset_diff: AssetDiffs,
}

/// Response for `wallet_prepareCalls`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PrepareCallsResponse {
    /// The [`PrepareCallsContext`] of the prepared call bundle.
    pub context: PrepareCallsContext,
    /// Digest of the prepared call bundle for the user to sign over
    /// with an authorized key.
    pub digest: B256,
    /// EIP-712 payload corresponding to the digest.
    pub typed_data: TypedData,
    /// Capabilities response.
    pub capabilities: PrepareCallsResponseCapabilities,
    /// Key that will be used to sign the call bundle.
    #[serde(default)]
    pub key: Option<CallKey>,
}

/// Response context from `wallet_prepareCalls`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum PrepareCallsContext {
    /// The [`SignedQuotes`] of the prepared call bundle.
    #[serde(rename = "quote")]
    Quote(Box<SignedQuotes>),
    /// The [`PreCall`] of the prepared call bundle.
    #[serde(rename = "preCall")]
    PreCall(SignedCall),
}

impl PrepareCallsContext {
    /// Initializes [`PrepareCallsContext`] with [`SignedQuotes`].
    pub fn with_quotes(quote: SignedQuotes) -> Self {
        Self::Quote(Box::new(quote))
    }

    /// Initializes [`PrepareCallsContext`] with a [`PreCall`].
    pub fn with_precall(precall: SignedCall) -> Self {
        Self::PreCall(precall)
    }

    /// Returns quotes immutable reference if it exists.
    pub fn quote(&self) -> Option<&SignedQuotes> {
        match self {
            PrepareCallsContext::Quote(signed) => Some(signed),
            PrepareCallsContext::PreCall(_) => None,
        }
    }

    /// Returns quotes mutable reference if it exists.
    pub fn quote_mut(&mut self) -> Option<&mut SignedQuotes> {
        match self {
            PrepareCallsContext::Quote(signed) => Some(signed),
            PrepareCallsContext::PreCall(_) => None,
        }
    }

    /// Consumes self and returns quotes if it exists.
    pub fn take_quote(self) -> Option<SignedQuotes> {
        match self {
            PrepareCallsContext::Quote(signed) => Some(*signed),
            PrepareCallsContext::PreCall(_) => None,
        }
    }

    /// Consumes self and returns precall if it exists.
    pub fn take_precall(self) -> Option<SignedCall> {
        match self {
            PrepareCallsContext::Quote(_) => None,
            PrepareCallsContext::PreCall(precall) => Some(precall),
        }
    }

    /// Calculate the digest that the user will need to sign.
    ///
    /// It will be a eip712 signing hash in a single chain intent and a merkle root in a multi chain
    /// intent.
    pub async fn compute_signing_digest(
        &self,
        orchestrator_address: Address,
        provider: &DynProvider,
    ) -> eyre::Result<(B256, TypedData)> {
        match self {
            PrepareCallsContext::Quote(context) => {
                let output_quote = context.ty().quotes.last().expect("should exist");
                if let Some(root) = context.ty().multi_chain_root {
                    Ok((root, TypedData::from_struct(&output_quote.output, None)))
                } else {
                    output_quote.output.compute_eip712_data(orchestrator_address, provider).await
                }
            }
            PrepareCallsContext::PreCall(pre_call) => {
                pre_call.compute_eip712_data(orchestrator_address, provider).await
            }
        }
    }
}

/// Capabilities for `wallet_sendPreparedCalls` request.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SendPreparedCallsCapabilities {
    /// Fee payment signature.
    #[serde(default)]
    pub fee_signature: Bytes,
}

/// Request parameters for `wallet_sendPreparedCalls`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendPreparedCallsParameters {
    /// The [`SendPreparedCallsCapabilities`] of the prepared call bundle.
    #[serde(default)]
    pub capabilities: SendPreparedCallsCapabilities,
    /// The [`PrepareCallsContext`] of the prepared call bundle.
    pub context: PrepareCallsContext,
    /// Key that was used to sign the call bundle.
    pub key: CallKey,
    /// Signature value.
    pub signature: Bytes,
}

/// Response for `wallet_sendPreparedCalls`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendPreparedCallsResponse {
    /// Bundle ID.
    pub id: BundleId,
}

/// The status code of a call bundle.
#[derive(Debug, Clone, Serialize_repr, Deserialize_repr, Eq, PartialEq)]
#[repr(u16)]
pub enum CallStatusCode {
    /// The call bundle is pending.
    Pending = 100,
    /// The call bundle was confirmed.
    Confirmed = 200,
    /// The call bundle failed offchain.
    Failed = 300,
    /// The call bundle reverted fully onchain.
    Reverted = 400,
    /// The call bundle partially reverted onchain.
    PartiallyReverted = 500,
}

impl CallStatusCode {
    /// Whether the bundle is pending.
    pub fn is_pending(&self) -> bool {
        matches!(self, CallStatusCode::Pending)
    }

    /// Whether the bundle status is final.
    pub fn is_final(&self) -> bool {
        !self.is_pending()
    }

    /// Whether the bundle was confirmed.
    pub fn is_confirmed(&self) -> bool {
        matches!(self, CallStatusCode::Confirmed)
    }

    /// Whether the bundle failed offchain.
    pub fn is_failed(&self) -> bool {
        matches!(self, CallStatusCode::Failed)
    }
}

/// A receipt for a call bundle.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallReceipt {
    /// The chain ID the transaction was included in.
    pub chain_id: ChainId,
    /// The logs generated in the transaction.
    pub logs: Vec<Log>,
    /// The status of the transaction.
    #[serde(flatten)]
    pub status: Eip658Value,
    /// The block hash the transaction was included in.
    pub block_hash: Option<BlockHash>,
    /// The block number the transaction was included in.
    pub block_number: Option<BlockNumber>,
    /// The gas used by the transaction.
    pub gas_used: u64,
    /// The transaction hash.
    pub transaction_hash: TxHash,
}

impl CallReceipt {
    /// Attempts to decode the logs to the provided log type.
    ///
    /// Returns the first log that decodes successfully.
    ///
    /// Returns None, if none of the logs could be decoded to the provided log type or if there
    /// are no logs.
    pub fn decoded_log<E: SolEvent>(&self) -> Option<alloy::primitives::Log<E>> {
        self.logs.iter().find_map(|log| E::decode_log(&log.inner).ok())
    }
}

/// The status of a call bundle.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CallsStatus {
    /// The ID of the call bundle.
    pub id: BundleId,
    /// The status of the call bundle.
    pub status: CallStatusCode,
    /// The receipts for the call bundle.
    pub receipts: Vec<CallReceipt>,
}
