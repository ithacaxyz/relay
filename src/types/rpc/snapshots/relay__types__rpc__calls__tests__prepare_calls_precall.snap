---
source: src/types/rpc/calls.rs
expression: response_value
---
{
  "capabilities": {
    "assetDiffs": {
      "0x0": [
        [
          "0x0000000000000000000000000000000000000000",
          [
            {
              "address": "0x0000000000000000000000000000000000000000",
              "decimals": 18,
              "direction": "incoming",
              "fiat": {
                "currency": "USD",
                "value": "0"
              },
              "name": "Ethereum",
              "recipients": [
                "0x0000000000000000000000000000000000000000"
              ],
              "symbol": "ETH",
              "type": "native",
              "uri": "https://ethereum.org",
              "value": "0x0"
            }
          ]
        ]
      ],
      "0x1": [
        [
          "0x0000000000000000000000000000000000000000",
          [
            {
              "address": "0x0000000000000000000000000000000000000000",
              "decimals": 18,
              "direction": "incoming",
              "fiat": {
                "currency": "GBP",
                "value": "0"
              },
              "name": "Ethereum",
              "recipients": [
                "0x0000000000000000000000000000000000000000"
              ],
              "symbol": "ETH",
              "type": "native",
              "uri": "https://ethereum.org",
              "value": "0x0"
            }
          ]
        ]
      ]
    },
    "authorizeKeys": [
      {
        "expiry": "0x0",
        "hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "permissions": [
          {
            "selector": "0x00000000",
            "to": "0x0000000000000000000000000000000000000000",
            "type": "call"
          },
          {
            "limit": "0x0",
            "period": "minute",
            "token": "0x0000000000000000000000000000000000000000",
            "type": "spend"
          }
        ],
        "publicKey": "0x0000000000000000000000000000000000000000",
        "role": "admin",
        "type": "secp256k1"
      }
    ],
    "feeTotals": {
      "0x0": {
        "currency": "USD",
        "value": "0"
      },
      "0x1": {
        "currency": "GBP",
        "value": "0"
      }
    },
    "revokeKeys": [
      {
        "hash": "0x0000000000000000000000000000000000000000000000000000000000000000"
      }
    ]
  },
  "context": {
    "preCall": {
      "chainId": "0x0",
      "eoa": "0x0000000000000000000000000000000000000000",
      "executionData": "0x",
      "nonce": "0x0",
      "signature": "0x"
    }
  },
  "digest": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "key": {
    "prehash": false,
    "publicKey": "0x0000000000000000000000000000000000000000",
    "type": "secp256k1"
  },
  "signature": "0xb6f20da7b95259dd00d3617be14d243171e09d6821f69a87557778b53d75b4e14c72b2dac47d4af67955a6b7010970a70cd2f5b4e1393dad0c5a9ba66435d8b01c",
  "typedData": {
    "domain": {},
    "message": null,
    "primaryType": "",
    "types": {}
  }
}
